### PR TITLE
[workloadmeta] Add ListContainersWithFilter func

### PR DESCRIPTION
**Pull request by davidor**
### What does this PR do?

This PR adds the possibility to use a filter when listing the containers stored in workloadmeta.

One of the filters defined allows the caller to receive only containers that are running. This helps us simplify some parts of the code that until now were listing all the containers and then discarding the ones that weren't running.

### Describe how to test/QA your changes

The features that rely on listing containers from the workloadmeta store should keep working as before and without reporting data from containers that are not running. Those features are: the container check, the process-agent and the security-agent. This PR modifies the same code as https://github.com/DataDog/datadog-agent/pull/12355 so they can be QA'd together.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/workloadmeta/store.go [hunk 1]
```diff
@@ -263,6 +263,24 @@ func (s *store) ListContainers() ([]*Container, error) {
...
+	var res []*Container
+
+	for _, container := range allContainers {
+		if filter(container) {
+			res = append(res, container)
+		}
+	}
+
+	return res, nil
+}
```
Comment (ahmed-mez) [inline pkg/workloadmeta/store.go:282]: I think we could make it more performant if we did the filtering in the `ListContainers()` loop [here](https://github.com/DataDog/datadog-agent/blob/3433b169377001bcb4d3e2ed0f86ceb128441e2d/pkg/workloadmeta/store.go#L258-L261)

File: pkg/workloadmeta/types.go [hunk 2]
```diff
@@ -57,6 +57,10 @@ type Store interface {
 	// to all entities with kind KindContainer.
 	ListContainers() ([]*Container, error)
```
Comment (vboulineau) [inline pkg/workloadmeta/types.go:58]: I don't think we need to keep that. Passing a nil filter should be enough

Comment (davidor) [reply pkg/workloadmeta/types.go:58]: Right, I thought about that, but for callers that just want to list everything it's easier than calling with something like `(nil)` or `(GetAll)`.

**Review by paulcacheux — APPROVED** [review]
LGTM, I added the `team/agent-security` label so that we get a QA card for this, since this is kind of critical. Thanks

**Review by ahmed-mez — APPROVED** [review]

**Comment by davidor** [comment]
I've rebased on top of `main` to fix conflicts with https://github.com/DataDog/datadog-agent/pull/12355

**Review by xornivore — APPROVED** [review]

**Review by vboulineau — APPROVED** [review]

**Pull request merged**